### PR TITLE
build(bazel): add bazel targets for aio doc generation

### PR DESCRIPTION
**Pull request by kormide**
@josephperrott @gkalpak @gregmagolan 

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [x] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the new behavior?

The bazel build for aio now outputs dgeni docs. This is an intermediary step in bazel migration. This breaks the legacy build and ci in this branch.

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No

### CODE REVIEW

File: aio/package.json [hunk 1]
```diff
@@ -19,7 +19,7 @@
     "build-local": "yarn ~~build",
     "prebuild-local-ci": "yarn setup-local-ci",
     "build-local-ci": "yarn ~~build --progress=false",
-    "extract-cli-command-docs": "node tools/transforms/cli-docs-package/extract-cli-commands.js a517160a2",
+    "extract-cli-command-docs": "node scripts/extract-cli-commands.js a517160a2",
```
Comment (kormide) [inline aio/package.json:22]: I moved this script out of transforms and into scripts, as it's done during setup rather than during the dgeni transform. It was also just easier bazel-wise as putting a `BUILD` file here would have made the content globbing more complicated.

File: aio/BUILD.bazel [hunk 2]
```diff
@@ -115,3 +115,53 @@ architect_test(
...
+        "@aio_npm//semver",
+        "@aio_npm//stemmer",
+        "@aio_npm//unist-util-filter",
+        "@aio_npm//unist-util-source",
+        "//:package.json",
+        "//:packages",
+        ":cli-src",
+    ],
+    env = {
+        "BAZEL_DGENI_OUTPUT_PATH": "$(@D)",
```
Comment (kormide) [inline aio/BUILD.bazel:163]: This env var is used to pass the bazel output path to the dgeni transform scripts so that they'll dump files in the output dir. You'll notice the conditional logic I added to those scripts that checks if we're in a bazel build. Those conditions are needed while we maintain both the legacy and bazel builds.

File: .bazelignore [hunk 3]
```diff
@@ -5,11 +5,100 @@
...
+# Ignore symlinked example node_modules directories. We need to list each 
+# path due to https://github.com/bazelbuild/bazel/issues/8106. We should be
+# able to remove this after flipping the build to bazel. To regenerate this list,
+# run the following command from the root directory:
+#
+#    find -L aio/content/examples -maxdepth 2 -type d -name node_modules | sort
+
+aio/content/examples/accessibility/node_modules
+aio/content/examples/ajs-quick-reference/node_modules
+aio/content/examples/angular-compiler-options/node_modules
```
Comment (kormide) [inline .bazelignore:22]: Unfortunate, but I don't think there's a cleaner way to do this.

Comment (kormide) [reply .bazelignore:22]: If I don't add these ignores, when I `glob` over the content folders bazel will examine every single node_modules folder causing a huge performance hit, even if they are added to an `exclude` list in glob. This is because bazel's glob evaluates all inclusions before removing the exclusions.

Comment (kormide) [reply .bazelignore:22]: We only need this because the legacy build symlinks the shared examples node_modules folder into each example in the source tree. After the flip, I can remove this list.

File: aio/tools/transforms/config.js [hunk 4]
```diff
@@ -1,14 +1,20 @@
...
+
 const PROJECT_ROOT = resolve(__dirname, '../../..');
 const AIO_PATH = resolve(PROJECT_ROOT, 'aio');
 const TEMPLATES_PATH = resolve(AIO_PATH, 'tools/transforms/templates');
 const API_TEMPLATES_PATH = resolve(TEMPLATES_PATH, 'api');
 const CONTENTS_PATH = resolve(AIO_PATH, 'content');
 const GUIDE_EXAMPLES_PATH = resolve(CONTENTS_PATH, 'examples');
 const SRC_PATH = resolve(AIO_PATH, 'src');
-const OUTPUT_PATH = resolve(SRC_PATH, 'generated');
+const OUTPUT_PATH = IS_BAZEL_BUILD ?  resolve(BAZEL_DGENI_PATH, 'generated') : resolve(SRC_PATH, 'generated');
```
Comment (kormide) [inline aio/tools/transforms/config.js:17]: I don't love how the term `OUTPUT_PATH` is hard to disambiguate with bazel output paths, but I was trying to change as little as possible. This actually refers to the dgeni output path.

File: aio/tools/transforms/angular-base-package/index.js [hunk 5]
```diff
@@ -157,6 +157,9 @@ module.exports = new Package('angular-base', [
 
   .config(function(postProcessHtml, addImageDimensions, autoLinkCode, filterPipes, filterAmbiguousDirectiveAliases, ignoreHttpInUrls, ignoreGenericWords) {
     addImageDimensions.basePath = path.resolve(AIO_PATH, 'src');
+    if (IS_BAZEL_BUILD) {
+      addImageDimensions.bazelOutPath = BAZEL_DGENI_PATH;
+    }
```
Comment (kormide) [inline aio/tools/transforms/angular-base-package/index.js:162]: "Why not just set basePath on a bazel build?", you ask. The issue is that some anchors still reference the source tree, even in a bazel build, so my solution was to try loading from the source tree then fall back to loading from the bazel out dir if not found.

File: aio/BUILD.bazel [hunk 6]
```diff
@@ -115,3 +115,53 @@ architect_test(
...
+    configuration_env_vars = ["PATH"],  # Need PATH because script invokes yarn
+    output_dir = True,
+    tool = "//aio/scripts:extract-cli-commands",
+)
+
+# Run dgeni generation
+npm_package_bin(
+    name = "dgeni",
+    args = ["./aio/tools/transforms/angular.io-package"],
+    configuration_env_vars = ["PATH"],  # Need PATH for dgeni package that invokes git
```
[the remark this answers is not in the document]
Comment (devversion) [reply aio/BUILD.bazel:135]: Also, is `git` just used for the version information? if so, we should add a comment that ideally we would just leverage stamping?

[the remark this answers is not in the document]
Comment (kormide) [reply aio/BUILD.bazel:135]: Some of the dgeni transforms make use of the dgeni git package for version information which calls git directly. Here I was tryingto change the dgeni packages as little as possible. I'll have to look deeper into what it's doing to see if we can use stamping instead.

File: aio/BUILD.bazel [hunk 7]
```diff
@@ -115,3 +115,53 @@ architect_test(
...
 )
+
+# Extract cli commands for dgeni generation
+npm_package_bin(
+    name = "cli-src",
+    args = [
+        "a517160a2",  # Keep tag in sync with package.json script 'extract-cli-command-docs' until flip
+        "$(@D)",
+    ],
+    configuration_env_vars = ["PATH"],  # Need PATH because script invokes yarn
```
[the remark this answers is not in the document]
Comment (devversion) [reply aio/BUILD.bazel:126]: I just read through the script (and now that we have a separate branch -- which is what I was raising yesterday exactly for allowing code like this to be more idiomatic): It seems that the AIO docs just rely on Yarn for convenience here, but really what we could do is:

1. Use an `http_archive` in the `WORKSPACE` for pulling the CLI snapshot repository
2. Leverage that in the Dgeni `cli-docs-package` (see `index.js`) -- using e.g. `@bazel/runfiles`

That way, we wouldn't need to rely on Yarn at all, and it would be idiomatic code -- with better caching as well

[the remark this answers is not in the document]
Comment (kormide) [reply aio/BUILD.bazel]: @devversion I updated WORKSPACE to load the cli sources and now the dgeni cli-docs-pacakge script points to those. I didn't need to use `@bazel/runfiles` because the dgeni generation is a build target and wouldn't access the cli source as runfiles.

[the remark this answers is not in the document]
Comment (devversion) [reply aio/BUILD.bazel]: Thanks, much better! 😄 (I think manifest/or CWD resolution would be a little more readable compared to going up to the execroot with many `../` segments -- but that is just a personal preference)

Also a question: Are you going to remove the `Is bazel` switches now that you can operate in the separate branch, not needing to support the standalone CLI build?

[the remark this answers is not in the document]
Comment (kormide) [reply aio/BUILD.bazel]: Yep I'll be removing those switches. Could you give me an example of what manifest resolution looks like? Still ramping up on bazel and haven't seen that before.

[the remark this answers is not in the document]
Comment (kormide) [reply aio/BUILD.bazel]: So correct me if I'm wrong, but I don't think that will work because the `npm_package_bin` dgeni target is a build target rather than an executable or test target, so including `@cli_src:files_for_docs` inside `data` like I'm doing [here](https://github.com/angular/angular/pull/45356/files#diff-b55a60abc10998e54981525fa4b55043fd1cf6a6c7438c3295e2026f55d13f50R146) doesn't make those files available as runfiles.

I could however try resolving from the cwd rather than the relative location from the index file.

[the remark this answers is not in the document]
Comment (devversion) [reply aio/BUILD.bazel]: oh yeah, you're right that there is neither a manifest in this case. There are cases where it seems like that & works as expected when e.g. a `nodejs_binary` executable with the runfiles is used as tool/executable for a build action.

I think `cwd`/exec-root-based paths are the most idiomatic approach here but I can see arguments against it since you never know if something in between changed the `CWD` here.

File: aio/tools/transforms/config.js [hunk 8]
```diff
@@ -1,14 +1,20 @@
 const { resolve } = require('path');
 const { readdirSync } = require('fs');
 
+// TODO: Remove non bazel build code paths after flip to bazel. Determine if
+// bazel build by presence of BAZEL_DGENI_OUTPUT_PATH env var.
+const IS_BAZEL_BUILD = !!process.env.BAZEL_DGENI_OUTPUT_PATH;
+const BAZEL_DGENI_PATH = process.env.BAZEL_DGENI_OUTPUT_PATH;
+const BAZEL_AIO_PATH = resolve(BAZEL_DGENI_PATH + '/../');
+
 const PROJECT_ROOT = resolve(__dirname, '../../..');
```
[the remark this answers is not in the document]
Comment (kormide) [reply aio/tools/transforms/config.js:5]: Since I didn't change anything here, I assume it does if Windows developers are building aio?

**Comment by kormide** [comment]
Updated to remove "is bazel checks" and stop supporting a working legacy build.

File: aio/tools/transforms/config.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (kormide) [reply aio/tools/transforms/config.js:5]: Note that the added line appearing in this diff is no longer there. Not sure why it still shows.

File: WORKSPACE [hunk 9]
```diff
@@ -29,9 +29,29 @@ http_archive(
...
+    name = "files_for_docgen",
+    srcs = glob([
+        "help/**/*.json",
+        "package.json",
+    ]),
+    visibility = ["//visibility:public"],
+)
+""",
+    sha256 = "7bdc978205985ff46a2ce30977c511e70798243d827854a85bf0bbd73e8621c9",
+    strip_prefix = "cli-builds-a517160a2",
```
Comment (gkalpak) [inline WORKSPACE:53]: Could we leverage `CLI_SRC_TAG` here too, so we don't have to update the hash in two places?

File: WORKSPACE [hunk 10]
```diff
@@ -29,9 +29,29 @@ http_archive(
...
+filegroup(
+    name = "files_for_docgen",
+    srcs = glob([
+        "help/**/*.json",
+        "package.json",
+    ]),
+    visibility = ["//visibility:public"],
+)
+""",
+    sha256 = "7bdc978205985ff46a2ce30977c511e70798243d827854a85bf0bbd73e8621c9",
```
Comment (gkalpak) [inline WORKSPACE:52]: What is this SHA computed from? (When do we need to update this SHA and what to?)

File: aio/package.json [hunk 11]
```diff
@@ -144,6 +144,7 @@
     "fs-extra": "^10.0.0",
     "globby": "^13.0.0",
     "hast-util-is-element": "^1.1.0",
+    "hast-util-has-property": "^1.0.4",
```
Comment (gkalpak) [inline aio/package.json:147]: OOC, why is this needed?

File: aio/tools/transforms/cli-docs-package/index.js [hunk 12]
```diff
@@ -4,8 +4,8 @@ const Package = require('dgeni').Package;
 const basePackage = require('../angular-base-package');
 const contentPackage = require('../content-package');
 const {CONTENTS_PATH, TEMPLATES_PATH, requireFolder} = require('../config');
-const CLI_SOURCE_ROOT = resolve(CONTENTS_PATH, 'cli-src');
-const CLI_SOURCE_PATH = resolve(CLI_SOURCE_ROOT, 'node_modules/@angular/cli');
+
+const CLI_SOURCE_PATH = resolve('../../../../../external/cli_src');
```
Comment (gkalpak) [inline aio/tools/transforms/cli-docs-package/index.js:8]: Where is `external/` located? Is it next to the root directory? (😱)

Also, could we use `PROJECT_ROOT` from [transforms/config.js][1] in the path to make it easier to follow?

[1]:https://github.com/angular/angular/blob/c0b2627565444eabbf2bb1c926a39a5a67c5e193/aio/tools/transforms/config.js#L5

File: aio/tools/transforms/angular-base-package/index.js [hunk 13]
```diff
@@ -136,7 +136,8 @@ module.exports = new Package('angular-base', [
     //  That being said do this only add 500ms onto the ~30sec doc-gen run - so not a huge issue)
     checkAnchorLinksProcessor.ignoredLinks.push({
       test(url) {
-        return (existsSync(resolve(SRC_PATH, url)));
+        // Some links point to assets in the source tree while others point to the generated bazel output
+        return existsSync(resolve(SRC_PATH, url)) || existsSync(resolve(BAZEL_OUTPUT_PATH, url));
```
Comment (gkalpak) [inline aio/tools/transforms/angular-base-package/index.js:140]: OOC, can you give an example of a link that points to the generated bazel output?

File: aio/tools/transforms/angular-base-package/index.js [hunk 14]
```diff
@@ -154,9 +155,12 @@ module.exports = new Package('angular-base', [
     ];
   })
 
-
   .config(function(postProcessHtml, addImageDimensions, autoLinkCode, filterPipes, filterAmbiguousDirectiveAliases, ignoreHttpInUrls, ignoreGenericWords) {
-    addImageDimensions.basePath = path.resolve(AIO_PATH, 'src');
+    // Some images exist within the source tree while others are in the generated bazel output
```
Comment (gkalpak) [inline aio/tools/transforms/angular-base-package/index.js:159]: Can you give an example of an image that is in the generated bazel output?

File: aio/BUILD.bazel [hunk 15]
```diff
@@ -115,3 +115,41 @@ architect_test(
...
 )
+
+# Run dgeni generation
+npm_package_bin(
+    name = "dgeni",
+    args = ["./aio/tools/transforms/angular.io-package"],
+    configuration_env_vars = ["PATH"],  # Need PATH for dgeni package that invokes git
+    data = glob([
+        "content/**",
+        "src/assets/**/*",
```
Comment (gkalpak) [inline aio/BUILD.bazel:126]: OOC, is the `/*` part necessary? Wouldn't `src/assets/**` work?

File: aio/BUILD.bazel [hunk 16]
```diff
@@ -115,3 +115,41 @@ architect_test(
     configuration_env_vars = ["NG_BUILD_CACHE"],
     data = TEST_FILES + TEST_DEPS,
 )
+
+# Run dgeni generation
+npm_package_bin(
+    name = "dgeni",
+    args = ["./aio/tools/transforms/angular.io-package"],
+    configuration_env_vars = ["PATH"],  # Need PATH for dgeni package that invokes git
+    data = glob([
```
Comment (gkalpak) [inline aio/BUILD.bazel:124]: Can you please add a comment on what things need to go into `data`? :pray:

File: aio/BUILD.bazel (hunk 6 above)
[the remark this answers is not in the document]
Comment (devversion) [reply aio/BUILD.bazel:123]: If this is something you want to do in a separate PR, a todo would be good I think

File: BUILD.bazel [hunk 17]
```diff
@@ -8,6 +8,7 @@ exports_files([
     "scripts/ci/payload-size.sh",
     "scripts/ci/payload-size.js",
     "package.json",
+    "packages",
```
Comment (devversion) [inline BUILD.bazel:11]: One question I have: Is this the long-term solution to getting access to the sources? This means that we often might invalidate the docs even if non-source files have changed right? In the components repo we have fine-grained `source-files` targets (definitely cumbersome -- but worth it I think).

File: WORKSPACE (hunk 10 above)
Comment (kormide) [reply WORKSPACE:52]: It's computed from the cli-builds tar. If you update `CLI_SRC_TAG` to a newer version and the hash is wrong, bazel will complain. I had to manually download the archive and compute the hash. I can add a comment with some commands to run.

File: aio/package.json (hunk 11 above)
Comment (kormide) [reply aio/package.json:147]: Hm, I could have sworn there was an error before and including this fixed it (maybe it was depended upon transitively?). But I can't reproduce it now so will remove.

File: BUILD.bazel (hunk 17 above)
Comment (kormide) [reply BUILD.bazel:11]: Good catch. Just an oversight on my part. I'll make the targets more fine-grained.

File: WORKSPACE (hunk 10 above)
Comment (gkalpak) [reply WORKSPACE:52]: Ah, that's not great, but I guess there is no way around it.
Yes, please, add any info that would be useful in updating this in the future 🙏

File: aio/tools/transforms/angular-base-package/index.js (hunk 13 above)
Comment (kormide) [reply aio/tools/transforms/angular-base-package/index.js:140]: `generated/images/marketing/concept-icons/animations.png` in `aio/contentmarketing/presskit.html`. The legacy build puts these under `src/generated` but all generated files in bazel are now under the bazel output directory.

File: aio/tools/transforms/angular-base-package/index.js (hunk 14 above)
Comment (kormide) [reply aio/tools/transforms/angular-base-package/index.js:159]: `generated/images/guide/architecture/metadata.png`. Same deal as the comment above.

File: WORKSPACE (hunk 9 above)
Comment (kormide) [reply WORKSPACE]: Fixed this.

File: WORKSPACE (hunk 10 above)
Comment (kormide) [reply WORKSPACE]: I added some more notes to this section and a command to run. Let me know if it's clear.

File: aio/tools/transforms/cli-docs-package/index.js (hunk 12 above)
Comment (kormide) [reply aio/tools/transforms/cli-docs-package/index.js]: It turns out I was escaping the sandbox to access the `external` folder in the execroot but I can just use the symlinked version inside the sandbox, so I could remove all of the `../`s.

To answer your question, `external` is a folder placed in the execroot (the cwd the bazel build runs in). It hosts any external repositories that are depended upon by the rule. In this case, we're depending on the cli-src tar we downloaded in `WORKSPACE`.

File: aio/BUILD.bazel (hunk 16 above)
Comment (kormide) [reply aio/BUILD.bazel]: I split this up into a couple of variables. Let me know if it's clearer now.

File: BUILD.bazel (hunk 17 above)
Comment (kormide) [reply BUILD.bazel:11]: Fixed. I added finer-grained targets throughout the angular project.

File: aio/BUILD.bazel (hunk 6 above)
[the remark this answers is not in the document]
Comment (kormide) [reply aio/BUILD.bazel:155]: As this PR is getting big, I would prefer to think about this separately. I added a note to our aio migration proposal doc.

**Comment by kormide** [comment]
Updated and ready for another review pass.

File: aio/content/cli-src/package.json [hunk 18]
```diff
@@ -0,0 +1,5 @@
+{
+  "dependencies": {
+    "@angular/cli": "https://github.com/angular/cli-builds#9ac3df5d1"
+  }
+}
```
Comment (kormide) [inline aio/content/cli-src/package.json:5]: TODO: Should be able to delete this.

File: aio/tools/transforms/angular-api-package/processors/processSpecialElements.js [hunk 19]
```diff
@@ -6,6 +6,7 @@ module.exports = function processSpecialElements() {
     $runBefore: ['collectPackageContentDocsProcessor'],
     $process(docs) {
       const moduleDocs = {};
+
```
Comment (kormide) [inline aio/tools/transforms/angular-api-package/processors/processSpecialElements.js:9]: TODO: remove space

File: packages/core/testing/BUILD.bazel [hunk 20]
```diff
@@ -2,7 +2,10 @@ load("//tools:defaults.bzl", "ng_module")
 
 package(default_visibility = ["//visibility:public"])
 
-exports_files(["package.json"])
+exports_files([
+    "package.json",
+    "index.ts",
```
Comment (gkalpak) [inline packages/core/testing/BUILD.bazel:7]: OOC, why do some sub-packages have to export `index.ts` and others don't?

File: WORKSPACE (hunk 10 above)
Comment (gkalpak) [reply WORKSPACE]: Looks great. Thx!

File: aio/BUILD.bazel (hunk 16 above)
Comment (gkalpak) [reply aio/BUILD.bazel]: :+1:

File: aio/tools/transforms/angular-base-package/index.js (hunk 13 above)
Comment (gkalpak) [reply aio/tools/transforms/angular-base-package/index.js:140]: I see, thx :+1:

File: aio/tools/transforms/angular-base-package/index.js (hunk 14 above)
Comment (gkalpak) [reply aio/tools/transforms/angular-base-package/index.js:159]: :+1:

File: packages/core/testing/BUILD.bazel (hunk 20 above)
Comment (kormide) [reply packages/core/testing/BUILD.bazel:7]: Oops, that was from a previous approach. The export allows you other targets to access files within this package. Instead of declaring the files as exports, I went with the `files_for_docgen` filegroup targets instead. Will delete.

**Comment by kormide** [comment]
Fixed up minor issues and TODOs.

File: BUILD.bazel (hunk 17 above)
Comment (devversion) [reply BUILD.bazel:11]: this is not needed anymore right?

File: packages/BUILD.bazel [hunk 21]
```diff
@@ -34,4 +34,65 @@ ts_config(
 exports_files([
     "license-banner.txt",
     "README.md",
+    "examples",
 ])
+
+# Run the following command to regenerate this list of all files_for_docgen targets.
+# bazel query 'attr("name", "files_for_docgen", //packages/... -//packages:files_for_docgen)' --output=label 2> /dev/null | sort | sed -e "s/\(.*\)/\"\1\",/"
```
Comment (devversion) [inline packages/BUILD.bazel:41]: You should have a look at how we enforce this always being up-to-date in the components repo. It's obviously much more work, but basically every package has a config of its entry-points, and that allows us to build up the list automatically here.

Obviously this is more work, and can be a clean-up, but can you add it to the doc for tracking? also potentially considering to use the new Bazel `subpackages` builtin

**Review by devversion — APPROVED** [review]

File: packages/BUILD.bazel (hunk 21 above)
Comment (kormide) [reply packages/BUILD.bazel:41]: Added a note to our shared doc to look into this.

**Review by gkalpak — APPROVED** [review]

**Review by josephperrott — APPROVED** [review]
LGTM

File: WORKSPACE [hunk 22]
```diff
@@ -32,9 +32,35 @@ http_archive(
...
+    url = "https://github.com/aspect-build/bazel-lib/archive/refs/tags/v0.5.3.tar.gz",
+)
+
+# Download cli source from angular/cli-builds for doc generation at a specific tag or commit.
+# If the ref is a commit sha, use the full sha instead of the abbreviated form. Tags, which
+# hold an abbreviated sha by convention in cli-builds, can continue to use the short form.
+CLI_SRC_REF = "c0a0bfb65df2cb2302131775517aa77c5c3161b9"
+
+http_archive(
+    name = "cli_src",
```
[the remark this answers is not in the document]
Comment (kormide) [reply WORKSPACE]: Updated the external repository name.
